helix-term/src/commands/typed.rs

```diff
@@ -1176,7 +1176,12 @@ fn reload(
     doc.reload(view, &cx.editor.diff_providers, redraw_handle)
         .map(|_| {
             view.ensure_cursor_in_view(doc, scrolloff);
-        })
+        })?;
+    if let Err(e) = doc.load_history() {
```

Comment:
On second thought, this would delete any changes not written.

Reply:
Yeah this shouldn't be there IMO, the history is modified by using a reload transaction, I think that should be respected in the (serialized) undo history.

Reply:
Removed

Reply:
I don't think this is a simple blaxk and white either load the history or don't. 

I think the we should always load the undofile into memory. Then any new revisions (simply check any revisions are newer than the last saved revision) would be merged into memory. 

The next step is to check if the undofile revision hash and the hash of the loaded file match. If that is the case we should just switch revisions to the head revision of the undofile. All edits happened in helix so no new transaction is needed.

If the two hashes don't match we compute the diff as we currently do and add a new transaction. 

This way no part of history (either in memory or in the file) ever gets lost, if all edits happen entirely in helix we just keep the history and switch the head revision (so doing edits like A-> B in instance 1 then B -> C in instance two and C -> D in instance one would form an history like A -> B -> C -> D)

Reply:
BTW. when overwriting a file with contention (meaning we write with `w!` because the file changed on disk we may also want to reload the history file and do a history merge as I described (but not update the revision ofcourse) so we don't accidentlly loose any history

Reply:
> The next step is to check if the undofile revision hash and the hash of the loaded file match.
Could you clarify what you mean by that?

Reply:
ah sorry forgot about that. Well the idea is the following:

If we are reloading and some random program that is not helix changed the file somehow then the right way to treat that is just a normal entry in the undo history (just like we do now).

However if the edits happened inside a second helix instance then we actually have a history of those edits and there is already a revision for them.
All we need to do is update the head revision to point to the head revision of the undofile. We don't need to generate a new transaction at all.

For example lets consider the following history:

We have an undofile:
```
A -> B (B is heaD)
```
And made some additional edits to that which we haven't saved yet:

````
A -> B -> C -> D (D is head)
````

Then we open that file in a second helix instance. This instance doesn't know about our new changes yet because we haven't saved.
We do some different edits in that second instance. It's history now looks like this.

```
A -> B -> E -> F (F is head)
```

Now we save in the second history so that history is written to the undofile.
Now if we reload in the first instance the following should happen:

* first we merge the histories

```
A -> B -> C -> D (D is head)
         \ 
           \ -> E -> F
```

* then we check if the hash stored in the undofile matches the file we just loaded (so all edits happened inside helix)
* in this case that is true (since we only edited in a second helix instance). That means we don't need to create a new revision for this edit because there is already a history for it. Simply move HEAD to `F`.
```
A -> B -> C -> D
         \ 
           \ -> E -> F (F is head)
```

That means that cooperation between two helix instances works seamlessly and two helix instances can share undo history. 

We can ofcourse only do this if the file at `F` is the same as the file on disk.
This would not be the case if another program (like `git reset --hard` ) changed the file between the save from helix and `:reload`. By comparing the SHA1 hashes we can avoid this problem.

Reply:
The merging would work as follows:

The Revision `B` appears both in the undofile and the in-memory version.
However there are new children in the undofile (old children can never be removed) . In this case `E`. Simply add these to the list of children of `B`.

Reply:
Or in other words: The merging would work by computing the union of child sets (if you want to get academic about it)